test/dataloader.jl

```diff
@@ -143,4 +143,27 @@
         d = DataLoader(x, batchsize=2, shuffle=true)
         @test length(collect(d)) == 3 # succesfull iteration
     end
+
+    @testset "Transducers foldl" begin
+        dloader = DataLoader(1:10)
+        @test foldl(+, Map(x -> x[1]), dloader; init = 0) == 55
+
+        dloader = DataLoader(1:10; shuffle = true)
+        @test foldl(+, Map(x -> x[1]), dloader; init = 0) == 55
+
+        dloader = DataLoader(1:10; batchsize = 2)
+        @test foldl(+, Map(x -> x[1]), dloader; init = 0) == 25
+
+        dloader = DataLoader(1:1000; shuffle = false)
+        @test copy(Map(x -> x[1]), Vector{Int}, dloader) == collect(1:1000)
+
+        dloader = DataLoader(1:1000; shuffle = true)
+        @test copy(Map(x -> x[1]), Vector{Int}, dloader) != collect(1:1000)
+
+        dloader = DataLoader(1:1000; batchsize = 2, shuffle = false)
+        @test copy(Map(x -> x[1]), Vector{Int}, dloader) == collect(1:2:1000)
+
+        dloader = DataLoader(1:1000; batchsize = 2, shuffle = true)
+        @test copy(Map(x -> x[1]), Vector{Int}, dloader) != collect(1:2:1000)
+    end
```

Comment:
So now the `@inferred` tests in this file are passing?

Reply:
They don't, I've not made any changes to the behavior of the iterator API. Good point on inference in general though, I changed `__foldl__` to infer correctly and added a test for it.